### PR TITLE
update: comet

### PR DESCRIPTION
**Pull request by MaLLLiYA**
## Description
Update tests
Add fragment ion indexing support

- [fragindex_max_fragmentmass](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_max_fragmentmass.html)
- [fragindex_min_fragmentmass](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_min_fragmentmass.html)
- [fragindex_min_ions_report](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_min_ions_report.html)
- [fragindex_min_ions_score](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_min_ions_score.html)
- [fragindex_num_spectrumpeaks](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_num_spectrumpeaks.html)
- [fragindex_skipreadprecursors](https://uwpr.github.io/Comet/parameters/parameters_202402/fragindex_skipreadprecursors.html)

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added support for Comet fragment ion indexing functionality with new parameters:
  - `fragindex_max_fragmentmass`
  - `fragindex_min_fragmentmass`
  - `fragindex_min_ions_report`
  - `fragindex_min_ions_score`
  - `fragindex_num_spectrumpeaks`
  - `fragindex_skipreadprecursors`
- Added custom residue mass configuration options
- Improved isotope error handling with expanded options
- Added protein delimiter configuration
- Updated test configurations and output files to match Comet 2024.02 rev. 0
- Updated various submodules (pyOpenMS docs, vcpkg, THIRDPARTY)

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CometAdapter.cpp</strong><dd><code>Add fragment ion indexing and residue mass customization</code>&nbsp; </dd></summary>
<hr>

src/topp/CometAdapter.cpp

<li>Added fragment ion indexing parameters and functionality<br> <li> Added support for custom residue mass settings<br> <li> Improved isotope error handling with expanded options<br> <li> Added protein delimiter configuration option<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-e1e99b4ea946a2cc81f22c595d19a1791202d46dccb4df3783971aa0c089e150">+43/-10</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CometAdapter_1.ini</strong><dd><code>Update test configuration with new Comet parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/topp/THIRDPARTY/CometAdapter_1.ini

<li>Added fragment ion indexing parameters<br> <li> Added residue mass customization options<br> <li> Updated isotope error restrictions<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-d3fbb7d27f2fe62878ec6b42918e63871256a2c732f8174eed07ca38470f553c">+34/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CometAdapter_3.ini</strong><dd><code>Update test configuration with new Comet parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/topp/THIRDPARTY/CometAdapter_3.ini

<li>Added fragment ion indexing parameters<br> <li> Added residue mass customization options<br> <li> Updated isotope error restrictions<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-f5795c01af2f551188d41f5b86b83f6792ea00bcf8c7292cc6f683ab0bb2cfc0">+34/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CometAdapter_2_out.idXML</strong><dd><code>Update test output for new Comet version</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/topp/THIRDPARTY/CometAdapter_2_out.idXML

<li>Updated test output file with new parameter values<br> <li> Modified search results to match new Comet version<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-15651109a58320b2e42883cbbdd86212fb3ce0a652e763a72d570f030d444586">+306/-273</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CometAdapter_3_out.idXML</strong><dd><code>Update test output for Comet fragment ion indexing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/topp/THIRDPARTY/CometAdapter_3_out.idXML

<li>Updated search engine version to 2024.02 rev. 0<br> <li> Added new fragment ion indexing parameters<br> <li> Modified score values and formatting for better precision<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-bbae5a2cd5e5517644042f309a0c96cfb5ea3763e3017b22c6b12f23ee77d157">+258/-225</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CometAdapter_1_out.idXML</strong><dd><code>Update test output for Comet fragment ion indexing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/topp/THIRDPARTY/CometAdapter_1_out.idXML

<li>Updated search engine version to 2024.02 rev. 0<br> <li> Added new fragment ion indexing parameters<br> <li> Adjusted score values and formatting<br> <li> Updated file paths to use Windows-style paths<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-9504aafcfd64525bb4d89bcc2ceded8383dc082ec9dc445750ecf65b9922ef86">+171/-138</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pyopenms-docs</strong><dd><code>Update pyOpenMS documentation submodule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/pyOpenMS/pyopenms-docs

- Updated submodule reference

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-4a6b2264479183bd9fba7178bc6ed1f9cfe484c1f12eb913e190abb759ee698a">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>vcpkg</strong><dd><code>Update vcpkg submodule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

vcpkg

- Updated submodule reference

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-ca4b9abdd99374b1e975427f9c9162f723d8125473a06ca29059630d347474b3">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>THIRDPARTY</strong><dd><code>Update THIRDPARTY submodule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

THIRDPARTY

- Updated submodule reference

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7715/files#diff-07f198312f299b3300d937248337f3a8f1885fdf747150ddbea12e44256a5838">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: src/topp/CometAdapter.cpp [hunk 1]
```diff
@@ -98,6 +98,14 @@ class TOPPCometAdapter :
 
   map<string,int> num_enzyme_termini {{"semi",1},{"fully",2},{"C-term unspecific", 8},{"N-term unspecific",9}};
 
+  std::string generateSetString_(const char residue) const
+  {
+    std::string set_str = "set_";
+    set_str += residue;
+    set_str += "_residue";
+    return set_str;
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp]:
```suggestion
  return "set_" + std::string(1, residue) + "_residue";
```

File: src/topp/CometAdapter.cpp [hunk 2]
```diff
@@ -210,6 +218,13 @@ class TOPPCometAdapter :
     registerIntOption_("spectrum_batch_size", "<posnum>", 20000, "max. number of spectra to search at a time; use 0 to search the entire scan range in one batch", false, true);
     setMinInt_("spectrum_batch_size", 0);
     registerDoubleList_("mass_offsets", "<doubleoffset1, doubleoffset2,...>", {0.0}, "One or more mass offsets to search (values subtracted from deconvoluted precursor mass). Has to include 0.0 if you want the default mass to be searched.", false, true);
+    registerStringOption_("pinfile_protein_delimiter", "<delimiter>", "", "specify a different character or string for the protein column delimiter (default tab)", false, true);
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp:220]: just an idea how to parameter description could be simplified
```suggestion
    registerStringOption_("pinfile_protein_delimiter", "<delimiter>", "\t", "protein column delimiter (default tab)", false, true);
```

File: src/topp/CometAdapter.cpp [hunk 3]
```diff
@@ -283,8 +317,10 @@ class TOPPCometAdapter :
     isotope_error["0/1"] = 1;
     isotope_error["0/1/2"] = 2;
     isotope_error["0/1/2/3"] = 3;
-    isotope_error["-8/-4/0/4/8"] = 4;
-    isotope_error["-1/0/1/2/3"] = 5;
+    isotope_error["-1/0/1/2/3"] = 4;
+    isotope_error["-1/0/1"] = 5;
+    isotope_error["-3/-2/-1/0/1/2/3"] = 6;
+    isotope_error["-8/-4/0/4/8"] = 7;
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp]: these strings are also used in the registerOptionsAndFlags_ for the isotope_error parameter.
Can we define them once and use there and here?

File: src/topp/CometAdapter.cpp [hunk 4]
```diff
@@ -319,7 +355,8 @@ class TOPPCometAdapter :
     os << "mass_type_parent = " << 1 << "\n";                    // 0=average masses, 1=monoisotopic masses
     os << "mass_type_fragment = " << 1 << "\n";                  // 0=average masses, 1=monoisotopic masses
     os << "precursor_tolerance_type = " << 1 << "\n";            // 0=MH+ (default), 1=precursor m/z; only valid for amu/mmu tolerances
-    os << "isotope_error = " << isotope_error[getStringOption_(Constants::UserParam::ISOTOPE_ERROR)] << "\n";                   // 0=off, 1=0/1 (C13 error), 2=0/1/2, 3=0/1/2/3, 4=-8/-4/0/4/8 (for +4/+8 labeling)
+    os << "isotope_error = " << isotope_error[getStringOption_(Constants::UserParam::ISOTOPE_ERROR)] << "\n";                   // 0=off, 1=0/1 (C13 error), 2=0/1/2, 3=0/1/2/3, 4=-1/0/1/2/3, 5=-1/0/1
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp]: here the comment doesn't seem to be up-to-date (e.g., only goes to 5)
maybe just remove?

File: src/topp/CometAdapter.cpp (hunk 3 above)
Comment (timosachsenberg) [reply src/topp/CometAdapter.cpp]: e.g., one could define a map from index to string and use that in both places?
I did not think it through but could work

File: src/topp/CometAdapter.cpp [hunk 5]
```diff
@@ -533,6 +570,13 @@ class TOPPCometAdapter :
     os << "output_suffix =\n";                                   // add a suffix to output base names i.e. suffix "-C" generates base-C.pep.xml from base.mzXML input
     os << "mass_offsets = " << ListUtils::concatenate(getDoubleList_("mass_offsets"), " ") << "\n"; // one or more mass offsets to search (values subtracted from deconvoluted precursor mass)
     os << "precursor_NL_ions =\n"; //  one or more precursor neutral loss masses, will be added to xcorr analysis 
+    os << "pinfile_protein_delimiter = " << getStringOption_("pinfile_protein_delimiter") << "\n";
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp:565]: would this also work with the way you did it before (empty string)?

File: src/topp/CometAdapter.cpp [hunk 6]
```diff
@@ -533,6 +570,13 @@ class TOPPCometAdapter :
...
+    for (char residue = 'A'; residue <= 'Z'; ++residue) {
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp:566]: does this make sense for the whole alphabet? Does comet support the whole alphabet (e.g. Z,B,J are not an amino acid but sometimes used for more than one ambiguous amino acids)

File: vcpkg
Comment (timosachsenberg) [inline vcpkg:1]: submodule update not needed in this PR

File: THIRDPARTY
Comment (timosachsenberg) [inline THIRDPARTY:1]: double check if we list the comet version in antoher file (e.g. not only this README.txt) in the THIRDPARTY repo. I think I recall that we listed it somewhere else...

**Review by timosachsenberg — CHANGES_REQUESTED** [review]
Thanks! Very nice PR. I have some minor comments and suggestions.
I will also add another reviewer.

File: src/tests/topp/THIRDPARTY/CometAdapter_1_out.idXML [hunk 7]
```diff
@@ -1,138 +1,171 @@
...
+				<UserParam type="string" name="MS:1002258" value="28"/>
+				<UserParam type="string" name="MS:1002259" value="52"/>
+				<UserParam type="string" name="num_matched_peptides" value="1"/>
+				<UserParam type="int" name="isotope_error" value="0"/>
+				<UserParam type="float" name="MS:1002252" value="3.734"/>
+				<UserParam type="float" name="COMET:xcorr" value="3.734"/>
+				<UserParam type="float" name="MS:1002253" value="0.0"/>
+				<UserParam type="float" name="COMET:deltaCn" value="0.0"/>
+				<UserParam type="float" name="MS:1002255" value="1096.0"/>
+				<UserParam type="float" name="COMET:spscore" value="1096.0"/>
```
Comment (cbielow) [inline src/tests/topp/THIRDPARTY/CometAdapter_1_out.idXML:129]: these scores changed slightly.
Is this expected?

File: src/topp/CometAdapter.cpp (hunk 3 above)
Comment (cbielow) [reply src/topp/CometAdapter.cpp]: > e.g., one could define a map from index to string and use that in both places?
> I did not think it through but could work 

yep, good idea.

File: src/topp/CometAdapter.cpp [hunk 8]
```diff
@@ -239,6 +254,18 @@ class TOPPCometAdapter :
...
+    registerIntOption_("fragindex_min_ions_report", "<num>", 3, "The minimum number fragment ions a peptide must match against the fragment ion index in order to report this peptide in the output", false, true);
+    setMinInt_("fragindex_min_ions_report", 1);
+    registerIntOption_("fragindex_min_ions_score", "<num>", 3, "The minimum number fragment ions a peptide must match against the fragment ion index in order to proceed to xcorr scoring", false, true);
+    setMinInt_("fragindex_min_ions_score", 1);
+    registerIntOption_("fragindex_num_spectrumpeaks", "<num>", 100, "The number of mass/intensity pairs that would be queried against the fragment ion index", false, true);
+    setMinInt_("fragindex_num_spectrumpeaks", 1);
+    registerStringOption_("fragindex_skipreadprecursors", "<bool>", "false", "Whether or not Comet reads all precursors from the input files, if true, skip reading precursors from the input file", false, true);
+    setValidStrings_("fragindex_skipreadprecursors", ListUtils::create<String>("true,false"));
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
    setValidStrings_("fragindex_skipreadprecursors",  {"true", "false"});
```

File: src/topp/CometAdapter.cpp [hunk 9]
```diff
@@ -239,6 +254,18 @@ class TOPPCometAdapter :
     registerStringOption_("require_variable_mod", "<bool>", "false", "If true, requires at least one variable modification per peptide", false, true);
     setValidStrings_("require_variable_mod", ListUtils::create<String>("true,false"));
 
+    // fragment ion indexing parameters
+    registerDoubleOption_("fragindex_max_fragmentmass", "<mass>", 2000.0, "The maximum fragment ion mass to include in the fragment ion index", false, true);
+    registerDoubleOption_("fragindex_min_fragmentmass", "<mass>", 200.0, "The minimum fragment ion mass to include in the fragment ion index", false, true);
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]: these could use a `setMinFloat_(...)` (>=0)

File: src/topp/CometAdapter.cpp (hunk 6 above)
Comment (cbielow) [reply src/topp/CometAdapter.cpp:566]: Comet does it like that https://uwpr.github.io/Comet/parameters/parameters_202402/set_X_residue.html
even though its not entirely clear what it means for the ambiguous AA's.

File: src/topp/CometAdapter.cpp [hunk 10]
```diff
@@ -533,6 +570,13 @@ class TOPPCometAdapter :
     os << "output_suffix =\n";                                   // add a suffix to output base names i.e. suffix "-C" generates base-C.pep.xml from base.mzXML input
     os << "mass_offsets = " << ListUtils::concatenate(getDoubleList_("mass_offsets"), " ") << "\n"; // one or more mass offsets to search (values subtracted from deconvoluted precursor mass)
     os << "precursor_NL_ions =\n"; //  one or more precursor neutral loss masses, will be added to xcorr analysis 
+    os << "pinfile_protein_delimiter = " << getStringOption_("pinfile_protein_delimiter") << "\n";
+    for (char residue = 'A'; residue <= 'Z'; ++residue) {
+      std::string set_str = generateSetString_(residue);
+      if (getDoubleOption_(set_str)) { 
+        os << set_str << " = " << getDoubleOption_(set_str) << "\n";
+      }
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
        os << set_str << " = " << getDoubleOption_(generateSetString_(residue)) << "\n";
```

Comment (cbielow) [reply src/topp/CometAdapter.cpp]: you can always write that out. Better for debugging and less code.

**Review by cbielow — CHANGES_REQUESTED** [review]
nice!
just a few things to consider.

File: src/topp/CometAdapter.cpp (hunk 4 above)
Comment (MaLLLiYA) [reply src/topp/CometAdapter.cpp]: removed

File: src/topp/CometAdapter.cpp (hunk 5 above)
Comment (MaLLLiYA) [reply src/topp/CometAdapter.cpp:565]: If this parameter is left blank or is missing, the default tab delimiter is used.

File: src/tests/topp/THIRDPARTY/CometAdapter_1_out.idXML (hunk 7 above)
Comment (timosachsenberg) [reply src/tests/topp/THIRDPARTY/CometAdapter_1_out.idXML:129]: From: https://github.com/UWPR/Comet/releases/tag/v2024.01.0
- Simplify the spectral processing for Sp scoring (preliminary score) by just taking the raw binned spectra and normalizing the max intensity to 100.
- 
is what I found

File: vcpkg
Comment (MaLLLiYA) [reply vcpkg:1]: Should I close this PR and create a new one? I don't seem to have modified anything in vcpkg. 
I got this message in this PR:
> This branch has conflicts that must be resolved
> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.
> Conflicting files
> vcpkg

Comment (timosachsenberg) [reply vcpkg:1]: I think this is not necessary. I can do it later

File: src/topp/CometAdapter.cpp [hunk 11]
```diff
@@ -130,7 +138,17 @@ class TOPPCometAdapter :
     //registerIntOption_("mass_type_fragment", "<num>", 1, "0=average masses, 1=monoisotopic masses", false, true);
     //registerIntOption_("precursor_tolerance_type", "<num>", 0, "0=average masses, 1=monoisotopic masses", false, false);
     registerStringOption_(Constants::UserParam::ISOTOPE_ERROR, "<choice>", "off", "This parameter controls whether the peptide_mass_tolerance takes into account possible isotope errors in the precursor mass measurement. Use -8/-4/0/4/8 only for SILAC.", false, false);
-    setValidStrings_(Constants::UserParam::ISOTOPE_ERROR, ListUtils::create<String>("off,0/1,0/1/2,0/1/2/3,-8/-4/0/4/8,-1/0/1/2/3"));
+    //setValidStrings_(Constants::UserParam::ISOTOPE_ERROR, ListUtils::create<String>("off,0/1,0/1/2,0/1/2/3,-1/0/1/2/3,-1/0/1,-3/-2/-1/0/1/2/3,-8/-4/0/4/8"));
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp]:
```suggestion
```

File: src/topp/CometAdapter.cpp [hunk 12]
```diff
@@ -272,19 +311,28 @@ class TOPPCometAdapter :
...
-    isotope_error["-8/-4/0/4/8"] = 4;
-    isotope_error["-1/0/1/2/3"] = 5;
+    isotope_error[isotope_error_str[0]] = 0;
+    isotope_error[isotope_error_str[1]] = 1;
+    isotope_error[isotope_error_str[2]] = 2;
+    isotope_error[isotope_error_str[3]] = 3;
+    isotope_error[isotope_error_str[4]] = 4;
+    isotope_error[isotope_error_str[5]] = 5;
+    isotope_error[isotope_error_str[6]] = 6;
+    isotope_error[isotope_error_str[7]] = 7;
```
Comment (timosachsenberg) [inline src/topp/CometAdapter.cpp]:
```suggestion
for (size_t i = 0; i < isotope_error_str.size(); ++i) 
{
    isotope_error[isotope_error_str[i]] = i;
}
```

**Review by timosachsenberg — APPROVED** [review]

**Comment by timosachsenberg** [comment]
I just read the https://uwpr.github.io/Comet/notes/20241001_FI.html
release notes.
As I understood it, it will only perform a fragment index search if an index is provided.
Can we add a test for that?

**Comment by MaLLLiYA** [comment]
> I just read the https://uwpr.github.io/Comet/notes/20241001_FI.html release notes. As I understood it, it will only perform a fragment index search if an index is provided. Can we add a test for that?

I will consider it later.

File: src/topp/CometAdapter.cpp [hunk 13]
```diff
@@ -98,6 +98,14 @@ class TOPPCometAdapter :
 
   map<string,int> num_enzyme_termini {{"semi",1},{"fully",2},{"C-term unspecific", 8},{"N-term unspecific",9}};
 
+  map<int, string> isotope_error_str {
+    {0, "off"}, {1, "0/1"}, {2, "0/1/2"}, {3, "0/1/2/3"}, {4, "-1/0/1/2/3"}, {5, "-1/0/1"}, {6, "-3/-2/-1/0/1/2/3"}, {7, "-8/-4/0/4/8"}};
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
  std::vector<String> isotope_error_str {"off", "0/1", "0/1/2", "0/1/2/3", "-1/0/1/2/3", "-1/0/1", "-3/-2/-1/0/1/2/3", "-8/-4/0/4/8"};
```

File: src/topp/CometAdapter.cpp [hunk 14]
```diff
@@ -130,7 +138,16 @@ class TOPPCometAdapter :
...
+    setValidStrings_(Constants::UserParam::ISOTOPE_ERROR, {
+                                                            isotope_error_str[0],
+                                                            isotope_error_str[1],
+                                                            isotope_error_str[2],
+                                                            isotope_error_str[3],
+                                                            isotope_error_str[4],
+                                                            isotope_error_str[5],
+                                                            isotope_error_str[6],
+                                                            isotope_error_str[7],
+                                                          });
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
    setValidStrings_(Constants::UserParam::ISOTOPE_ERROR, isotope_error_str);
```

File: src/topp/CometAdapter.cpp [hunk 15]
```diff
@@ -210,6 +227,13 @@ class TOPPCometAdapter :
     registerIntOption_("spectrum_batch_size", "<posnum>", 20000, "max. number of spectra to search at a time; use 0 to search the entire scan range in one batch", false, true);
     setMinInt_("spectrum_batch_size", 0);
     registerDoubleList_("mass_offsets", "<doubleoffset1, doubleoffset2,...>", {0.0}, "One or more mass offsets to search (values subtracted from deconvoluted precursor mass). Has to include 0.0 if you want the default mass to be searched.", false, true);
+    registerStringOption_("pinfile_protein_delimiter", "<delimiter>", "", "specify a different character or string for the protein column delimiter (default tab)", false, true);
+    
+    for (char residue = 'A'; residue <= 'Z'; ++residue)
+    {
+      std::string set_str = generateSetString_(residue);
+      registerDoubleOption_(set_str, "<mass>", 0.0, "Redefine the base mass of this residue; applies to both the average and monoisotopic mass (default of 0.0 will use Comet's default masses)", false, true);
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp:225]:
```suggestion
      registerDoubleOption_(set_str, "<mass>", 0.0, "Redefine the base mass of this residue; applies to both the average and monoisotopic mass (default of 0.0 will use Comet's default masses)", false, true);
      setMinFloat_(set_str, 0.0);
```

File: src/topp/CometAdapter.cpp [hunk 16]
```diff
@@ -272,19 +310,24 @@ class TOPPCometAdapter :
...
-    isotope_error["off"] = 0;
-    isotope_error["0/1"] = 1;
-    isotope_error["0/1/2"] = 2;
-    isotope_error["0/1/2/3"] = 3;
-    isotope_error["-8/-4/0/4/8"] = 4;
-    isotope_error["-1/0/1/2/3"] = 5;
+for (size_t i = 0; i < isotope_error_str.size(); ++i) 
+{
+    isotope_error[isotope_error_str[i]] = i;
+}
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
```

Comment (cbielow) [reply src/topp/CometAdapter.cpp]: remove `map<string,int> isotope_error;` as well. Not needed anymore

File: src/topp/CometAdapter.cpp [hunk 17]
```diff
@@ -319,7 +362,8 @@ class TOPPCometAdapter :
     os << "mass_type_parent = " << 1 << "\n";                    // 0=average masses, 1=monoisotopic masses
     os << "mass_type_fragment = " << 1 << "\n";                  // 0=average masses, 1=monoisotopic masses
     os << "precursor_tolerance_type = " << 1 << "\n";            // 0=MH+ (default), 1=precursor m/z; only valid for amu/mmu tolerances
-    os << "isotope_error = " << isotope_error[getStringOption_(Constants::UserParam::ISOTOPE_ERROR)] << "\n";                   // 0=off, 1=0/1 (C13 error), 2=0/1/2, 3=0/1/2/3, 4=-8/-4/0/4/8 (for +4/+8 labeling)
+    os << "isotope_error = " << isotope_error[getStringOption_(Constants::UserParam::ISOTOPE_ERROR)] << "\n";
```
Comment (cbielow) [inline src/topp/CometAdapter.cpp]:
```suggestion
    os << "isotope_error = " << std::distance(isotope_error.begin(), std::find(isotope_error.begin(), isotope_error.end(), getStringOption_(Constants::UserParam::ISOTOPE_ERROR))) << "\n";                   
```

**Review by cbielow — APPROVED** [review]
nice.

I've added a few suggestions which would switch from
std::map to std::vector using less code (which is usually a good thing)
Optional though.

File: src/topp/CometAdapter.cpp (hunk 17 above)
Comment (MaLLLiYA) [reply src/topp/CometAdapter.cpp]: This should probably be isotope_error_str.begin() and isotope_error_str.end(), right?

File: src/topp/CometAdapter.cpp (hunk 16 above)
Comment (MaLLLiYA) [reply src/topp/CometAdapter.cpp]: removed

**Comment by timosachsenberg** [comment]
looks good to me... I would suggest we wait a bit with the merge until we get to do proper benchmarking using quantms

**Pull request closed without merging**